### PR TITLE
Time trend BoxCox transform and weight trending

### LINKED ISSUE #477: Add time trend in risk factors
We are adding an underlying time trend in risk factors from consuming UPF (ultra-processed food).
The plan is to add time (or year) as an additional variable into risk factor equations.

- [x] The parameters need to be estimated.
- [x] It needs to be confirmed how this time trend will affect other equations in Health-GPS.
- [x] #492 
- [x] #493

### LINKED ISSUE #499: Add boxcox params for time trends
We need to use a BoxCox transform for the time trend values. This will require loading in extra params:

* `ExpectedTrendBoxCox`, different from `ExpectedTrend`
* `Lambda` for `Trend` models

we also need to do the BoxCox transform itself in `update_trends`.

### LINKED ISSUE #501: Apply time trend to body weight
Body weight also needs time trending.

Mean weight adjustment happens after nutrients intake is adjusted with time trend.

For a certain age, gender group,
mean weight = 0.06256 * mean time-trend-adjusted-Energy Intake - 0.6256 * mean height + 0.4925 * age - 16.6166 * gender + 16.1161

where mean time-trend-adjusted-Energy Intake = 4 * mean time-trend-adjusted-carb + 9 * mean time-trend-adjusted-fat + 4 * mean time-trend-adjusted-protein + 0 * mean time-trend-adjusted-sodium

### PR DESCRIPTION
**Pull request by jamesturner246**
Resolves #499 
Resolves #501 
Resolves #477

The main changes here are:
* Implement BoxCox transformation on time trend factors
* Implement recursive `get_expected` in `KevinHallModel`, to compute expected weight from nutrients
* Remove default arguments to `RiskFactorAdjustableModel` methods to pre-empt bugs
* Added a static gender to value method to `person.cpp`

Happy to walk through over a chat.

### CODE REVIEW

File: src/HealthGPS/kevin_hall_model.cpp [hunk 1]
```diff
@@ -258,14 +258,58 @@ KevinHallModel::compute_weight_adjustments(RuntimeContext &context,
...
+    if (factor == "EnergyIntake"_id) {
+        double energy_intake = 0.0;
+        for (const auto &[nutrient_key, energy_coefficient] : energy_equation_) {
+            double nutrient_intake =
+                get_expected(context, sex, age, nutrient_key, std::nullopt, apply_trend);
+            energy_intake += nutrient_intake * energy_coefficient;
+        }
+        return energy_intake;
+    }
+
```
Comment (jzhu20) [inline src/HealthGPS/kevin_hall_model.cpp:297]: Is this energy intake with time trend? Sorry I cannot tell.

**Comment by jzhu20** [comment]
It looks good to me.

File: src/HealthGPS.Input/model_parser.cpp [hunk 2]
```diff
@@ -187,7 +187,9 @@ load_staticlinear_risk_model_definition(const nlohmann::json &opt, const Configu
     std::vector<core::DoubleInterval> policy_ranges;
     auto trend_models = std::make_unique<std::vector<LinearModelParams>>();
     auto trend_ranges = std::make_unique<std::vector<core::DoubleInterval>>();
+    auto trend_lambda = std::make_unique<std::vector<double>>();
     auto expected_trend = std::make_unique<std::unordered_map<core::Identifier, double>>();
+    auto expected_trend_boxcox = std::make_unique<std::unordered_map<core::Identifier, double>>();
```
Comment (alexdewar) [inline src/HealthGPS.Input/model_parser.cpp:192]: I know we talked about this on the last PR... but I'm still a bit confused as to why these need to be `std::unique_ptr`s. Even if they're being converted into `std::shared_ptr`s at some point can't you just `std::move` your `std::vector` (for example) into a `std::shared_ptr<std::vector<T>>`?

File: src/HealthGPS/kevin_hall_model.cpp [hunk 3]
```diff
@@ -258,14 +258,58 @@ KevinHallModel::compute_weight_adjustments(RuntimeContext &context,
...
+    // Compute expected body weight.
+    if (factor == "Weight"_id) {
+        double weight = 16.1161;
+        weight +=
+            0.06256 * get_expected(context, sex, age, "EnergyIntake"_id, std::nullopt, apply_trend);
+        weight -= 0.6256 * get_expected(context, sex, age, "Height"_id, std::nullopt, false);
+        weight += 0.4925 * age;
+        weight -= 16.6166 * Person::gender_to_value(sex);
+        return weight;
```
Comment (alexdewar) [inline src/HealthGPS/kevin_hall_model.cpp:306]: Where do these coefficients come from?

File: src/HealthGPS/person.cpp [hunk 4]
```diff
@@ -47,7 +47,9 @@ double Person::get_risk_factor_value(const core::Identifier &key) const {
     throw std::out_of_range("Risk factor not found: " + key.to_string());
 }
 
-float Person::gender_to_value() const {
+float Person::gender_to_value() const { return gender_to_value(gender); }
```
Comment (alexdewar) [inline src/HealthGPS/person.cpp:50]: I'm wondering whether these helpers should be member functions of the `core::Gender` `enum class` instead?

**Review by alexdewar — COMMENTED** [review]
I've made a few small comments... I'd be interested to have a walkthrough though if you've got time!

File: src/HealthGPS/kevin_hall_model.cpp (hunk 3 above)
Comment (jamesturner246) [reply src/HealthGPS/kevin_hall_model.cpp:306]: @jzhu20?

**Comment by jamesturner246** [comment]
> I've made a few small comments... I'd be interested to have a walkthrough though if you've got time!

Sure! Today is good.

File: src/HealthGPS/person.cpp (hunk 4 above)
Comment (jamesturner246) [reply src/HealthGPS/person.cpp:50]: Good shout.

File: src/HealthGPS.Input/model_parser.cpp (hunk 2 above)
Comment (jamesturner246) [reply src/HealthGPS.Input/model_parser.cpp:192]: Inside model parser, probably. I'll have a look.

**Comment by alexdewar** [comment]
> Sure! Today is good.

Sorry, only just saw this... Could do tomorrow?

**Pull request merged**